### DISCUSSION TITLE
OPTUNA Trials

### DISCUSSION THREAD

**Discussion by kanzaali** [optuna/optuna#3998] [Q&A]
### What is an issue?

Hi, 
Thank you so much for your work. 
I have some questions about OPTUNA-Trials, what it is doing exactly and why it is so time-consuming, I have three different types of objective functions and I am running each of them for 50-OPTUNA Trials. I want to know the insight for each OPTUNA Trial for each objective.  

`Objective 1
optuna_initial_lr = optuna_trial.suggest_float('LearningRate-for-Random-ImageNet', 1e-3, 1e-1, log=True)

Objective 2
optim_epochs = optuna_trial.suggest_int('no_epochs', 100, 300, step = 50) 

Objective 3
optuna_name = optuna_trial.suggest_categorical('optimizer', ['Adam', 'SGD', 'RMSprop', 'SAMSGD']) 

                if optuna_name == 'Adam':
                        criterion      = nn.CrossEntropyLoss(reduction='none')
                        optim_backbone = optim.Adam(models.parameters(), weight_decay=WDECAY)
                if optuna_name == 'SGD':
                        criterion      = nn.CrossEntropyLoss(reduction='none')
                        optim_backbone = optim.SGD(models.parameters(), lr=LR, momentum=MOMENTUM, weight_decay=WDECAY)
                if optuna_name == 'RMSprop':
                        criterion      = nn.CrossEntropyLoss(reduction='none')
                        optim_backbone = optim.RMSprop(models.parameters(), momentum=MOMENTUM, weight_decay=WDECAY)
                if optuna_name == 'SAMSGD':
                        criterion      = nn.CrossEntropyLoss(reduction='mean')
                        base_optimizer = torch.optim.SGD
                        optim_backbone = SAM(models.parameters(), base_optimizer, lr=LR, momentum=MOMENTUM, weight_decay=WDECAY)

  
    import optuna
    from optuna_custom_visualization import *
    import matplotlib.pyplot as plt
    

    study = optuna.create_study(direction="maximize")  # Create a new study.
    study.optimize(Objective(), n_trials=50)  # Invoke optimization of the objective function.

    #Slice Plot
    
    plot_slice = optuna.visualization.plot_slice(study,  target_name = 'Test Accuracy')
    plot_slice.write_image("SlicePlot-Random-CIFAR100-OptimalOptimizer.png", width = 1635, height = 923)

                
`

**Comment by nzw0301** [comment]
Sorry, I couldn't understand the question. Could you elaborate a little bit more about your question?

**Comment by kanzaali** [comment]
When I run one trial of OPTUNA for let's suppose objective 2 for finding the optimal epochs for my method it takes 

optim_epochs = optuna_trial.suggest_int('no_epochs', 100, 300, step = 50)

238 minutes for a single OPTUNA Trial _With 1 GPU and Memory_ 
238 × 50 OPTUNA Trials  ×  6 Different Epochs= 71400 Minutes or 49.58 days (50 Trials of OPTUNA will take )

Though I am running my work on 4GPUs OPTUNA still has not finished finding optimal epochs. 

Is my understanding wrong about OPTUNA trials or my method of using OPTUNA is wrong?

**Reply by nzw0301** [reply]
Usually, Optuna doesn't become a bottleneck in the running time. In your case, I suppose your model (neural net?) takes too much time to train. https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/003_efficient_optimization_algorithms.html#pruning-algorithms might be useful for your case that stops unpromising trials before the end of the training.

**Reply by kanzaali** [reply]
Thank you so much for your reply!
This link is not working BTW

**Reply by nzw0301** [reply]
Thank you for letting me know that! I've fixed the link.

**Comment by kanzaali** [comment]
Hi again, I have returned Accuracy for finding the optimal optimizer, I am unable to understand the values on Y-Axis. I am not getting which optimizer is best. 
![Slice-Random-CIFAR100-OptimalOptimizer](https://user-images.githubusercontent.com/33715985/190288417-f559b8bf-68e9-4779-a904-d169b233d699.png)
